src/Workspaces/Core/Portable/Extensions/ControlFlowGraphExtensions.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+    internal static partial class ControlFlowGraphExtensions
+    {
+        public static BasicBlock EntryBlock(this ControlFlowGraph cfg)
+        {
+            var firstBlock = cfg.Blocks[0];
+            Debug.Assert(firstBlock.Kind == BasicBlockKind.Entry);
+            return firstBlock;
+        }
+
+        public static BasicBlock ExitBlock(this ControlFlowGraph cfg)
+        {
+            var lastBlock = cfg.Blocks[cfg.Blocks.Length - 1];
+            Debug.Assert(lastBlock.Kind == BasicBlockKind.Exit);
+            return lastBlock;
+        }
+
+        public static IEnumerable<IOperation> DescendantOperations(this ControlFlowGraph cfg)
+        {
+            foreach (BasicBlock block in cfg.Blocks)
+            {
+                foreach (IOperation operation in block.DescendantOperations())
+                {
+                    yield return operation;
+                }
+            }
+        }
+
+        public static IEnumerable<T> DescendantOperations<T>(this ControlFlowGraph cfg, OperationKind operationKind)
+            where T : IOperation
+        {
+            Debug.Assert(cfg != null);
+
+            foreach (var descendant in cfg.DescendantOperations())
+            {
+                if (descendant?.Kind == operationKind)
+                {
+                    yield return (T)descendant;
+                }
```

Comment:
isn't this just  `cfg.DesendentOperations().Where(d => d?.Kind == operationKind)`?

Reply:
Yes, that is correct. I preferred to write it this way to keep it consistent with BasicBlockExtensions, which needs separate loops to deal with operations and branch value.

---
In reply to: [230606563](https://github.com/dotnet/roslyn/pull/30777#discussion_r230606563) [](ancestors = 230606563)